File: trunk/data/modules/cargo_mission.py
import universe
import random
import launch
import faction_ships
import VS
import Briefing
class cargo_mission:
	you=VS.Unit()
	faction=""
	destination=""
	base=VS.Unit()
	cargoname=""
	arrived=0
	difficulty=1
	distfrombas=0
	quantity=1
	cred=0
	capship=0
	mission_time=0
	jumps=()
        mplay=self.mplay
	def initbriefing(self):
		self.jump_ani=0
		self.rnd_y=0.0
		self.added_warp=1
		self.brief_stage=0
		self.begintime= VS.getGameTime()-6.0
		print "starting briefing"
		if (self.you.isNull()):
			VS.terminateMission(0)
			Briefing.terminate()
			return
		faction=you.getFaction()
		name=you.getName()
		self.brief_you=Briefing.addShip(name,faction,(0.0,0.0,80.0))
		VS.IOmessage (0,"cargo mission","briefing","Your mission for today will be to deliver some %s cargo to the %s system.\nIn order to get there, you must follow this route that we have planned out for you." % (cargoname,destination))
	
	def loopbriefing(self):
		size=len(jumps)
		time = VS.getGameTime()
		Briefing.setCamPosition((1.6*(time-self.begintime)*self.brief_stage,0.0,0.0))
		if (((time-self.begintime)>=5.0) and added_warp):
			self.jump_ani=Briefing.addShip("brief_warp",faction,(20.0*(brief_stage),rnd_y,79.5+rnd_y))
			self.added_warp=0
		if (((time-self.begintime)>=6.0)):
			if (self.jump_ani!=0):
				Briefing.removeShip(self.jump_ani)
				self.jump_ani=0
		if ((size==self.brief_stage) and ((time-self.begintime)>=6.0)):
			VS.IOmessage(0,"cargo mission","briefing","Once there, you must drop the cargo off at a specified unit")
			self.brief_stage=size+1
			self.added_warp=0
			self.time=0.0
		elif ((self.brief_stage>size) and ((time-self.begintime)>=11.0)):
			Briefing.terminate()
			return
		elif (((time-self.begintime)>=6.0) and (self.brief_stage<size)):
			self.added_warp=1
			self.rnd_y=(random.random()*40.0-20.0)
			Briefing.addShip("brief_jump",faction,(20.0*(self.brief_stage+1),self.rnd_y,79.6+self.rnd_y))
			Briefing.enqueueOrder (self.brief_you,(20.0*(self.brief_stage+1),self.rnd_y,80.0+self.rnd_y,5.0))
			self.begintime=time
			myname=self.jumps[self.brief_stage]
			VS.IOmessage (0,"cargo mission","briefing","You must go to the '%s' jump point" % (myname))
			self.brief_stage+=1
	
	def endbriefing(self):
		print "endinging briefing"
		del self.jump_ani
		del self.rnd_y
		del self.added_warp
		del self.brief_stage
		del self.begintime
	
	def __init__ (self,factionname, numsystemsaway, cargoquantity, missiondifficulty, distance_from_base, creds, launchoncapship, time_to_complete, category):
	  self.mission_time=VS.getGameTime()+time_to_complete*100*float(1+numsystemsaway)
	  self.capship= launchoncapship
	  self.faction=factionname
	  self.cred=creds
#####	  universe.init()
	  self.distfrombase=distance_from_base
	  self.difficulty=missiondifficulty
	  mysys=VS.getSystemFile()
	  self.quantity=cargoquantity
	  sysfile = mysys
	  self.you=VS.getPlayer()
	  mplay=universe.getMessagePlayer(self.you)
	  if (self.quantity<1):
	    self.quantity=1
	  carg=VS.getRandCargo(quantity,category)
	  if (carg.GetQuantity()==0):
	    carg = VS.getRandCargo(quantity,"")
	  tempquantity=self.quantity
	  self.cargoname=carg.GetContent()
	  name = you.getName ()
	  if (self.you):
	    quantity = self.you.addCargo(carg)  #I add some cargo
	  else:
	    VS.terminateMission (0)
	    return
	  creds_deducted = (carg.GetPrice()*float(self.quantity)*random.random()+1)
	  self.cred += creds_deducted
	  if (tempquantity>0):
	    self.cred*=float(quantity)/float(tempquantity)
	  else:
	    VS.IOmessage (2,"cargo mission",self.mplay,"You do not have space to add our cargo to the mission. Mission failed.")
	    VS.terminateMission(0)
	  return
	  
	  if (quantity==0):
	    VS.IOmessage (2,"cargo mission",self.mplay,"You do not have space to add our cargo to the mission. Mission failed.")
	    VS.terminateMission(0)
 	    return
	  
	  VS.IOmessage (0,"cargo mission",self.mplay,"Good Day, %s. Your mission is as follows:" % (name))
	  (self.destination,self.jumps)=universe.getAdjacentSystems(sysfile,numsystemsaway)
	  VS.IOmessage (2,"cargo mission",self.mplay,"and give the cargo to a %s unit." % (faction))
	  VS.IOmessage (3,"cargo mission",self.mplay,"You will receive %d of the %s cargo" % (quantity,cargoname))
	  VS.IOmessage (4,"cargo mission",self.mplay,"We will deduct %.2f credits from your account for the cargo needed." % (creds_deducted))
	  VS.IOmessage (5,"cargo mission",self.mplay,"You will earn %.2f more credits when you deliver our cargo." % (creds))
	  self.you.addCredits (-creds_deducted)
	
	def takeCargoAndTerminate (self,you, remove):
	  removenum=0 #if you terminate without remove, you are SKREWED
	  if (remove):
	    removenum=you.removeCargo(cargoname,quantity,1)
	  
	  if ((removenum==self.quantity) or (self.quantity==0)):
	    VS.IOmessage (0,"cargo mission",self.mplay,"Excellent work pilot.")
	    VS.IOmessage (0,"cargo mission",self.mplay,"You have been rewarded for your effort as agreed.")
	    VS.IOmessage (0,"cargo mission",self.mplay,"Your excellent work will be remembered.")
	    you.addCredits(self.cred)
	    VS.terminateMission(1)
	    return
	  else:
	    VS.IOmessage (0,"cargo mission",self.mplay,"You did not follow through on your end of the deal.")
	    if (self.difficulty<1):
	      VS.IOmessage (0,"cargo mission",self.mplay,"Your pay will be reduced")
	      VS.IOmessage (0,"cargo mission",self.mplay,"And we will consider if we will accept you on future missions.")
	      addcred=(float(removenum)/float((self.quantity*(1+self.difficulty))))*self.cred
	      VS.addCredits(you,addcred)
	    else:
	      VS.IOmessage (0,"cargo mission",self.mplay,"You will not be paid!")
	      if (self.difficulty>=2):
		VS.IOmessage (0,"cargo mission",self.mplay,"And your idiocy will be punished.")
		VS.IOmessage (0,"cargo mission",self.mplay,"You had better run for what little life you have left.")
		for i in range(self.difficulty):
		  un=faction_ships.getRandomFighter(self.faction)
		  newunit=launch.launch_wave_around_unit("shadow", self.faction, un, "default", 1, 200.0,400.0,you)
		  newunit.setFgDirective("B")
		  newunit.setTarget(you)
	    VS.terminateMission(0)
	    return
	  
	
	def Execute (self):
##	  if (VS.getGameTime()>mission_time):
##	    VS.IOmessage (0,"cargo mission",self.mplay,"You Have failed to deliver your cargo in a timely manner.")
##	    VS.IOmessage (0,"cargo mission",self.mplay,"The cargo is no longer of need to us.")
##	    if (you):
##	      takeCargoAndTerminate(you,0)
##	    return
	  if (arrived):
	    if (self.base.isNull() or self.you.isNull()):
	      VS.IOmessage (0,"cargo mission",self.mplay,"Mission failed. You were unable to deliver cargo.")
	      VS.terminateMission(0)
	      return
	    dist=you.getSignificantDistance(base)
	    if (dist<=self.distfrombase):
	      self.takeCargoAndTerminate(self.you,1)		
	      return
	  else:
	    sysfil = VS.getSystemFile()
	    if (sysfil==destination):
	      self.arrived=1
	      newship=faction_ships.getRandomCapitol(faction)
	      randint=random.randrange(0,50)
	      significant = unit.getSignificant (randint,(not capship),0)
	      if (_std.isNull (significant)):
		significant =VS.getPlayer()
	      if (_std.isNull(significant)):
		self.arrived=0
	      else:
		newun=significant
		if (self.capship):
		  newun=launch.launch_wave_around_unit("Base",self.faction,newship,"sitting_duck",1,2000.0,5000.0,significant)
		name = newun.getName ()
		VS.IOmessage (0,"cargo mission",self.mplay,"You must drop your cargo off with the %s." % (name))
		if (capship):
		  name=significant.getName()
		  VS.IOmessage (0,"cargo mission",self.mplay,"It is docked around the %s landmark." % (name))
		self.base=newun

def initrandom (factionname, missiondifficulty,creds_per_jump, launchoncapship, sysmin, sysmax, time_to_complete, category):
	return cargo_mission(factionname,random.random(sysmin,sysmax), random.randrange(4,15), missiondifficulty,400.0,creds_per_jump*float(1+numsys),launchoncapship, 10.0, category)

File: trunk/data/modules/cargo_mission.c
module cargo_mission {
  import ai_stationary;
  import universe;
  object youcontainer;
	object faction;
	object destination;
	object basecontainer;
	object cargoname;
	bool arrived;
	int difficulty;
	float distfrombase;
	int quantity;
	float cred;
	import random;
	import launch;
	import faction_ships;
	bool capship;
	float mission_time;

	float begintime;
	float time;
	bool done;
	void initbriefing() {
	  done=false;
	  begintime= _std.getGameTime();
	  _io.printf ("starting briefing");
	  int a=_briefing.addShip("starrunner","confed",0.0,0.0,100.0);

	  // int b=_briefing.addShip("starrunner","confed",0.0,0.0,1000.0);
	  //int c=_briefing.addShip("starrunner","confed",0.0,0.0,10000.0);
	  //int d=_briefing.addShip("starrunner","confed",100.0,0.0,00.0);
	  _briefing.enqueueOrder (a,200.0,10.0,1000.0,15.0);
	  //_briefing.enqueueOrder (b,4000.0,0.0-10.0,2000.0,15.0);
	  //_briefing.enqueueOrder (c,1000.0,10.0,1000.0,15.0);
	  //_briefing.enqueueOrder (d,4000.0,0.0-10.0,2000.0,15.0);
	};
	void loopbriefing() {
	  
	  time = _std.getGameTime();
	  _io.printf ("running briefing %f", time-begintime);
	  if ((time-begintime)<15.0){
	    object str = _string.new();
	    _io.sprintf (str,"%f",time);
	    _io.message (0,"game","briefing",str);
	    _string.delete(str);
	  }else {
	    if (!done) {
	      done=true;
	      _io.message (0,"game","briefing","WE HAVE FINISHED MOVING");	  
	    }

	  }
	  /*
	  if ((time-begintime)>6.0) {
	    _briefing.setCloak (0,0.5);//1-(time-begintime)/6.0);	 
	    }*/
	  if ((time-begintime)>20.0) {
	    _io.printf ("DADADATERMINATED");
	    _briefing.terminate();
	  }
	};
	void endbriefing() {
	  _io.printf ("endinging briefing");
	};
	void initrandom (object factionname, int missiondifficulty,float creds_per_jump, bool launchoncapship, int sysmin, int sysmax, float time_to_complete, object category) {
	  int numsys = random.randomint (sysmin,sysmax);
	  init(factionname,numsys, random.randomint(4,15), missiondifficulty,400.0,creds_per_jump*_std.Float(1+numsys),launchoncapship, 10.0, category);
	};
	void init (object factionname, int numsystemsaway, int cargoquantity, int missiondifficulty, float distance_from_base, float creds, bool launchoncapship, float time_to_complete, object category) {
	  mission_time=_std.getGameTime()+time_to_complete*_std.Float(1+numsystemsaway);
	  _std.setNull (youcontainer);
	  _std.setNull(basecontainer);
	  capship= launchoncapship;
	  faction_ships.init();
	  faction=_string.new();
	  _io.sprintf (faction,"%s",factionname);
	  arrived=false;
	  cred=creds;
	  distfrombase=distance_from_base;
	  difficulty=missiondifficulty;
	  object mysys=_std.getSystemFile();
	  quantity=cargoquantity;
	  object sysfile = _std.getSystemFile();
	  object you=_unit.getPlayer();
	  youcontainer=_unit.getContainer (you);
	  if (quantity<1){
	    quantity=1;
	  }
	  object list=_unit.getRandCargo(quantity,category);
	  while ((_std.Int(_olist.at(list,5))>5)) {
	    _olist.delete (list);
	    list = _unit.getRandCargo(quantity);
	  }
	  int tempquantity=quantity;
	  cargoname=_olist.at(list,0);
	  object str = _string.new();
	  if (!_std.isNull(you)) {
	    quantity = _unit.addCargo(you,cargoname        ,_olist.at(list,1),_olist.at(list,2),_olist.at(list,3),_olist.at(list,4),_olist.at(list,5));  //ADD CARGO HERE
	    object name = _unit.getName (you);
	    _io.sprintf(str,"Good Day, %s. Your mission is as follows:",name);
	    _string.delete (name);
	  } else {
	    //don't destroy--somethign else went wrong
	    _std.terminateMission (false);
	    return;
	  }
	  if (tempquantity>0) {
	    cred=cred *_std.Float(quantity)/_std.Float(tempquantity);
	  }
	  _io.message (0,"game","all",str);
	  destination=universe.getAdjacentSystem(sysfile,numsystemsaway);
	  _io.sprintf(str,"and give the cargo to a %s unit.",faction);
	  _io.message (2,"game","all",str);
	  _io.sprintf(str,"You will receive %d of the %s cargo",quantity,cargoname);
	  _io.message (3,"game","all",str);
	  _string.delete(str);
	  _olist.delete(list);
	};
	void destroy() {
	  _unit.deleteContainer (youcontainer);
	  if (arrived) {
	    if (!_std.isNull(basecontainer)) {
	      _unit.deleteContainer (basecontainer);
	    }
	  }
	  _string.delete (destination);
	  _string.delete (faction);
	};
	void takeCargoAndTerminate (object you, bool remove) {
	  int removenum=0; //if you terminate without remove, you are SKREWED
	  if (remove) {
	    removenum=_unit.removeCargo(you,cargoname,quantity,true);
	  }
	  if ((removenum==quantity)||(quantity==0)) {
	    _io.message (0,"game","all","Excellent work pilot.");
	    _io.message (0,"game","all","You have been rewarded for your effort as agreed.");
	    _io.message (0,"game","all","Your excellent work will be remembered.");
	    _unit.addCredits(you,cred);
	    destroy();
	    _std.terminateMission(true);
	    return;
	  } else {
	    _io.message (0,"game","all","You did not follow through on your end of the deal.");

	    if (difficulty<1) {
	      _io.message (0,"game","all","Your pay will be reduced");
	      _io.message (0,"game","all","And we will consider if we will accept you on future missions.");
	      float addcred=(_std.Float(removenum)/_std.Float((quantity*(1+difficulty))))*cred;
	      _unit.addCredits(you,addcred);
	    } else {
	      _io.message (0,"game","all","You will not be paid!");
	      if (difficulty>=2) {
		_io.message (0,"game","all","And your idiocy will be punished.");
		_io.message (0,"game","all","You had better run for what little life you have left.");
		int i=0;
		object un;


		while (i<difficulty) {
		  un=faction_ships.getRandomFighter(faction);
		  object newunit=launch.launch_wave_around_unit("shadow", faction, un, "default", 1, 200.0,400.0,you);
		  _unit.setTarget(newunit,you);
		  i=i+1;
		}
	      }
	    }
	    destroy();
	    _std.terminateMission(false);
	    return;
	  }
	};
	void loop () {
	  if (_std.getGameTime()>mission_time) {
	    _io.message (0,"game","all","You Have failed to deliver your cargo in a timely manner.");
	    _io.message (0,"game","all","The cargo is no longer of need to us.");
	    object you = _unit.getUnitFromContainer (youcontainer);
	    if (!_std.isNull(you)) {
	      takeCargoAndTerminate(you,false);
	    }
	    return;
	  }
	  if (arrived) {
	    object base=_unit.getUnitFromContainer(basecontainer);
	    object you=_unit.getUnitFromContainer(youcontainer);
	    if (_std.isNull(base)||_std.isNull(you)) {
	      _io.message (0,"game","all","Mission failed. You were unable to deliver cargo.");
	      destroy();
	      _std.terminateMission(false);
	      return;
	    }
	    float dist=_unit.getDistance(base,you);
	    if (dist<=distfrombase) {
	      takeCargoAndTerminate(you,true);		
	      return;
	    }
	  } else {
	    object sysfil = _std.getSystemFile();
	    if (_string.equal (sysfil,destination)) {
	      arrived=true;
	      object newship=faction_ships.getRandomCapitol(faction);
	      int randint=random.randomint(0,50);
	      object significant = unit.getSignificant (randint,(!capship));
	      if (_std.isNull (significant)) {
		significant =_unit.getPlayer();
	      }
	      if (_std.isNull(significant)) {
		arrived=false;
	      }else {
		object newun=significant;
		if (capship) {
		  newun=launch.launch_wave_around_unit("Base",faction,newship,"_ai_stationary",1,2000.0,5000.0,significant);
		}
		object str = _string.new();
		object name = _unit.getName (newun);
		_io.sprintf(str,"You must drop your cargo off with the %s.",name);
		_string.delete (name);
		_io.message (0,"game","all",str);
		if (capship) {
		  name=_unit.getName(significant);
		  _io.sprintf(str,"It is docked around the %s landmark.",name);
		  _string.delete (name);
		  _io.message (0,"game","all",str);
		  
		}
		_string.delete(str);
		
		basecontainer=_unit.getContainer(newun);
	      }
	    }
	    _string.delete (sysfil);
	  }
	};
}
